cfi_intro.adoc

```diff
@@ -43,247 +44,163 @@ address of the instruction following the jump (`pc+2`) to the link register `x1`
-entry point for shared functions, with each entry point having its own landing
-pad instruction followed by a direct branch to the start of the function. This
-would allow the function to be labeled with multiple labels, each corresponding
-to a specific call site.
-
-A portion of the label space may be dedicated to labeled landing pads that are
-only valid targets of an indirect jump (and not an indirect call).
-
-====
-
-Forward-edge and backward-edge CFI may be activated independently for software
-that executes in U-mode, S-mode, or M-mode. The processor keeps track of the CFI
-activation and CFI state for each mode in the `mstatus` CSR. A subset of the
-fields in the `mstatus` CSR are accessible using the `sstatus` CSR. VS-mode’s
-version of `sstatus` (`vsstatus`) tracks the CFI state for VS-mode and VU-mode.
-
-[NOTE]
-====
-The operating system may activate the use of Zicfisslp by U-mode applications,
-with or without the extension being used by the operating system itself. The set
-of U-mode programs installed in an OS may be a mix, where some programs are
-compiled with Zicfisslp capabilities and others that are not. The operating
-system can activate or deactivate the use of the extension in U-mode per
-application by context switching the Zicfisslp state.
-
-Hypervisors may activate the use of Zicfisslp in a virtual machine, with or
-without the extension being used by the hypervisor itself. Virtual machines that
-use the extension may coexist with virtual machines that do not, with the
-hypervisor context switching the Zicfisslp state of each virtual machine.
-
-To use Zicfisslp, the operating system and hypervisors have to be modified to
-use the capabilities, including the context switching of the extension state.
-
-Machine mode firmware may activate the use of the extension in M-mode independent
-of its use in lower privilege modes.
-====
-
-The Zicfisslp shadow stack instructions are encoded using a subset of "May be op"
-instructions defined by the Zimop and Zcmop extensions. This subset of
-instructions revert to their Zimop/Zcmop defined behavior when the Zicfisslp
-extension is not implemented or if the extension has not been activated at a
-privilege mode. A program that is built with Zicfisslp instructions can thus
-continue to operate correctly, but without control-flow integrity, on processors
-that do not support the extension or if the extension is not active. The
-Zicfisslp landing pad instruction uses the code points defined as HINTs for the
-`AUIPC` opcode. When forward-edge CFI is not active at a privilege level or when
-the extension is not implemented, the landing pad instruction executes as a
-no-op.
-
-[NOTE]
-====
 Compilers should flag each object file (for example, using flags in the elf
-attributes) to indicate if the object file has been compiled with the Zicfisslp
+attributes) to indicate if the object file has been compiled with the Zicfiss
```

Comment:
This needs to be more than a binary flag since mixing object files that are known to require xBCFIE=0 (conflicting use of Zimop encodings) with object files that are known to require xBCFIE=1 (control stack mode) must fail.  Precisely speaking, I think there are up to four properties that are ANDed by static and dynamic linking:

1. ZICFISS_AWARE: true if a zicfiss-supporting compiler is used, or if explicitly marked in assembly code.  If false, no assumptions can be made about stack discipline in the combined object.
2. ZICFISS_ALL_FUNCTIONS_PROTECTED: true if zicfiss is enabled in the compiler or if explicitly marked in assembly code.  If false, some functions are potentially unprotected.  This has no runtime effect but is useful for auditors.
3. PERMIT_BCFIE_0: false if control stack mode is used, or if the object file otherwise assumes that BCFIE=1 and a valid shadow stack is installed.
4. PERMIT_BCFIE_1: false if the object file assumes that a shadow stack is _not_ used, for instance an unwinder or coroutine library that does not provide support for shadow stacks.

Any link which causes PERMIT_BCFIE_0 and PERMIT_BCFIE_1 to both be false, fails.

We may choose to simplify the attributes by making ZICFISS_AWARE imply PERMIT_BCFIE_1 and ZICFISS_ALL_FUNCTIONS_PROTECTED and not explicitly tracking the latter two properties (this would mean that object files that are known to be incompatible with or incompletely protected by shadow stacks are erroneously marked as "unknown", but such cases are likely to be rare compared to true unknowns.)

Reply:
Here the text is calling out need for one or more flags but is not attempting to detail the flags. I was thinking there being two flags - Zicfiss_shadow and Zicfiss_control -  both assume all functions are protected. Zicfiss_shadow being portable to platforms where Zicfiss  is not implemented  or  not active. Zicfiss_control requires the extension to be active. Additional flags may be introduced if there is a need that arises.